### PR TITLE
gRPC api v2: Transaction Status

### PR DESCRIPTION
**Pull request by Bargsteen**
## Purpose

Add the transaction status endpoint to the new gRPC v2 API.

## Changes

- Add a large number of types needed for the transaction status.
   - As discussed with Ales, we use the more sane representation of trx status as defined in the rust sdk.

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: v2/concordium/types.proto [hunk 1]
```diff
@@ -78,6 +83,14 @@ message Release {
   repeated TransactionHash transactions = 3;
 }
 
+// A new individual release. Part of a single transfer with schedule transaction.
+message NewRelease {
+  // Effective time of the release in milliseconds since unix epoch.
+  uint64 timestamp = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:89]: We have a type for timestamp
```suggestion
  Timestamp timestamp = 1;
```

File: v2/concordium/types.proto [hunk 2]
```diff
@@ -583,11 +596,48 @@ message InitName {
...
 
+// Status of a transaction known to the node.
+message TransactionStatus {
+
+  message Committed {
+    repeated TransactionSummaryInBlock outcomes = 1;
+  }
+
+  message Finalized {
+    TransactionSummaryInBlock outcome = 1; // TODO: internalservererror grpc error if there are multiple
```
Comment (limemloh) [inline v2/concordium/types.proto:617]: Is the todo implemented?

**Review by limemloh — APPROVED** [review]
Only one comment

File: v2/concordium/types.proto [hunk 3]
```diff
@@ -607,6 +657,1084 @@ message Duration {
   uint64 value = 1;
 }
 
+// A reason for why a transaction was rejected. Rejected means included in a
+// block, but the desired action was not achieved. The only effect of a
+// rejected transaction is payment.
+message RejectReason {
+
+  message ModuleHashAlreadyExists {
+    ModuleRef contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:666]:
```suggestion
    ModuleRef module = 1;
```

File: v2/concordium/types.proto [hunk 4]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message InvalidAccountReference {
+    AccountAddress contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:670]:
```suggestion
    AccountAddress address = 1;
```

File: v2/concordium/types.proto [hunk 5]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+    InitName init_name = 2;
+  }
+
+  message InvalidReceiveMethod {
+    ModuleRef module_ref = 1;
+    ReceiveName receive_name = 2;
+  }
+
+  message InvalidModuleReference {
+    ModuleRef contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:684]:
```suggestion
    ModuleRef module_ref = 1;
```

File: v2/concordium/types.proto [hunk 6]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message InvalidContractAddress {
+    ContractAddress contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:688]:
```suggestion
    ContractAddress address = 1;
```

File: v2/concordium/types.proto [hunk 7]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+    ReceiveName receive_name = 3;
+    Parameter parameter = 4;
+  }
+
+  message AlreadyABaker {
+    BakerId contents = 1;
+  }
+
+  message NotABaker {
+    AccountAddress contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:712]:
```suggestion
    AccountAddress address = 1;
```

File: v2/concordium/types.proto [hunk 8]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message DuplicateAggregationKey {
+    BakerAggregationVerifyKey contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:716]:
```suggestion
    BakerAggregationVerifyKey key = 1;
```

File: v2/concordium/types.proto [hunk 9]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message EncryptedAmountSelfTransfer {
+    AccountAddress contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:720]:
```suggestion
    AccountAddress address = 1;
```

File: v2/concordium/types.proto [hunk 10]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message ScheduledSelfTransfer {
+    AccountAddress contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:724]:
```suggestion
    AccountAddress address = 1;
```

File: v2/concordium/types.proto [hunk 11]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message DuplicateCredIds {
+    repeated CredentialRegistrationId contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:728]:
```suggestion
    repeated CredentialRegistrationId ids = 1;
```

File: v2/concordium/types.proto [hunk 12]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message NonExistentCredIds {
+    repeated CredentialRegistrationId contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:732]:
```suggestion
    repeated CredentialRegistrationId ids = 1;
```

File: v2/concordium/types.proto [hunk 13]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message NotADelegator {
+    AccountAddress contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:736]:
```suggestion
    AccountAddress address = 1;
```

File: v2/concordium/types.proto [hunk 14]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+
+  message DelegationTargetNotABaker {
+    BakerId contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:740]:
```suggestion
    BakerId baker_id = 1;
```

File: v2/concordium/types.proto [hunk 15]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+
+message ContractTraceElement {
+  // A contract transferred an amount to an account.
+  message Transferred {
+    // Sender contract.
+    ContractAddress from = 1;
+    // Amount transferred.
+    Amount amount = 2;
+    // Receiver account.
+    AccountAddress to_ = 3;
```
Comment (limemloh) [inline v2/concordium/types.proto:931]:
```suggestion
    AccountAddress receiver = 3;
```

File: v2/concordium/types.proto [hunk 16]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  ReceiveName receive_name = 6;
+  // Any contract events that might have been generated by the contract execution.
+  repeated ContractEvent events = 7;
+}
+
+message ContractTraceElement {
+  // A contract transferred an amount to an account.
+  message Transferred {
+    // Sender contract.
+    ContractAddress from = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:927]:
```suggestion
    ContractAddress sender = 1;
```

File: v2/concordium/types.proto [hunk 17]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+    AccountAddress to_ = 2;
+    // Memo.
+    optional Memo memo = 3;
+  }
+  // An account was deregistered as a baker. This is the result of a
+  // successful UpdateBakerStake transaction.
+  message BakerStakeUpdated {
+    // If the stake was updated (that is, it changed and did not stay the
+    // same) then this is present, otherwise it is not present.
+    optional BakerStakeUpdatedData contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:1197]:
```suggestion
    optional BakerStakeUpdatedData update = 1;
```

File: v2/concordium/types.proto [hunk 18]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+    repeated CredentialRegistrationId new_cred_ids = 1;
+    // The credentials that were removed.
+    repeated CredentialRegistrationId removed_cred_ids = 2;
+    // The (possibly) updated account threshold.
+    AccountThreshold new_threshold = 3;
+  }
+  // A baker was configured. The details of what happened are contained in
+  // the list of BakerEvents.
+  message BakerConfigured {
+    repeated BakerEvent contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:1235]:
```suggestion
    repeated BakerEvent events = 1;
```

File: v2/concordium/types.proto [hunk 19]
```diff
@@ -607,6 +657,1084 @@ message Duration {
...
+  }
+  // An account configured delegation. The details of what happened are
+  // contained in the list of DelegationEvents.
+  message DelegationConfigured {
+    repeated DelegationEvent contents = 1;
```
Comment (limemloh) [inline v2/concordium/types.proto:1240]:
```suggestion
    repeated DelegationEvent events = 1;
```

**Review by limemloh — COMMENTED** [review]
Found a bit more

File: v2/concordium/types.proto (hunk 3 above)
Comment (Bargsteen) [reply v2/concordium/types.proto:666]: 😈

File: v2/concordium/types.proto (hunk 4 above)
Comment (Bargsteen) [reply v2/concordium/types.proto:670]: I actually just removed several of these simple wrapper types as I deemed them unnecessary.

File: v2/concordium/types.proto [hunk 20]
```diff
@@ -607,6 +657,1040 @@ message Duration {
...
+  ContractAddress address = 3;
+  // The amount the instance was initialized with.
+  Amount amount = 4;
+  // The name of the contract.
+  InitName init_name = 5;
+  // Any contract events that might have been genereated by the contract initialization.
+  repeated ContractEvent events = 6;
+}
+
+// Events that come from inside a smart contract.
```
Comment (abizjak) [inline v2/concordium/types.proto:853]:
```suggestion
// An event generated by a smart contract.
```

File: v2/concordium/types.proto [hunk 21]
```diff
@@ -607,6 +657,1040 @@ message Duration {
...
+  Amount amount = 4;
+  // The parameter passed to the method.
+  Parameter parameter = 5;
+  // The name of the method that was executed.
+  ReceiveName receive_name = 6;
+  // Any contract events that might have been generated by the contract execution.
+  repeated ContractEvent events = 7;
+}
+
+message ContractTraceElement {
```
Comment (abizjak) [inline v2/concordium/types.proto:879]: Documentation?

File: v2/concordium/types.proto [hunk 22]
```diff
@@ -607,6 +657,1040 @@ message Duration {
...
+    // The delegator's delegation target was updated.
+    DelegationSetDelegationTarget delegation_set_delegation_target = 4;
+    // A delegator was added.
+    DelegatorId delegation_added = 5;
+    // A delegator was removed.
+    DelegatorId delegation_removed = 6;
+  }
+}
+
+message AccountTransactionEffects {
```
Comment (abizjak) [inline v2/concordium/types.proto:1123]: There should be a short comment explaining the purpose of this type.

File: v2/concordium/types.proto [hunk 23]
```diff
@@ -607,6 +657,1040 @@ message Duration {
...
+}
+
+// A minting rate of CCD.
+message MintRate {
+  uint32 mantissa = 1;
+  // This will never exceed 255 and can thus be stored in a single byte.
+  uint32 exponent = 2;
+}
+
+message UpdatePayload {
```
Comment (abizjak) [inline v2/concordium/types.proto:1447]: Documentation.

**Review by abizjak — APPROVED** [review]
Looks very good.

File: v2/concordium/types.proto [hunk 24]
```diff
@@ -583,11 +596,48 @@ message InitName {
...
+  bytes value = 1;
+}
+
 // A smart contract v0 state.
 message ContractStateV0 {
   bytes value = 1;
 }
 
+// Status of a transaction known to the node.
+message TransactionStatus {
```
Comment (abizjak) [inline v2/concordium/types.proto:610]: Maybe we could name this `BlockItemStatus` instead?

Below there is a type called "TransactionSummaryInBlock", which is used here, but then its main field is called `BlockItemSummary` which is just a tiny bit confusing.

File: v2/concordium/service.proto [hunk 25]
```diff
@@ -53,4 +53,7 @@ service Queries {
   // Get information about the current state of consensus.
   rpc GetConsensusInfo (Empty) returns (ConsensusInfo);
 
+  // Get the status of and information about a specific transaction.
+  rpc GetTransactionStatus (TransactionHash) returns (TransactionStatus);
```
Comment (abizjak) [inline v2/concordium/service.proto:57]: And we could name this `GetBlockItemStatus` instead?

**Pull request merged**